


<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>API reference &#8212; Osprey 3.0-beta1 documentation</title>
    <link rel="stylesheet" href="_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.0-beta1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Tutorial" href="tutorial.html" />
  
   

  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial.html" title="Tutorial"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Osprey 3.0-beta1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
        <a href="
    index.html" class="text-logo">Osprey 3.0-beta1 documentation</a>
        
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Table Of Contents</h2>
  </div>
  <div class="sidebar-toc">
    
    
      <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API reference</a></li>
</ul>

    
  </div>
</div>
        
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="index.html">Docs</a></li>
              
              <li>API reference</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <div class="section" id="api-reference">
<span id="id1"></span><h1>API reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h1>
<p>To use Osprey in your Python script, import the <code class="docutils literal"><span class="pre">osprey</span></code> module
and run the <a class="reference internal" href="#osprey.start" title="osprey.start"><code class="xref py py-meth docutils literal"><span class="pre">osprey.start()</span></code></a> function:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">osprey</span>
<span class="n">osprey</span><span class="o">.</span><span class="n">start</span><span class="p">()</span>
</pre></div>
</div>
<dl class="function">
<dt id="osprey.start">
<code class="descclassname">osprey.</code><code class="descname">start</code><span class="sig-paren">(</span><em>heapSizeMiB=1024</em>, <em>enableAssertions=False</em>, <em>stackSizeMiB=16</em>, <em>garbageSizeMiB=128</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.start" title="Permalink to this definition">¶</a></dt>
<dd><p>Starts the Java Virtual Machine (JVM) that runs Osprey’s computation libraries.</p>
<p>Call <a class="reference internal" href="#osprey.start" title="osprey.start"><code class="xref py py-meth docutils literal"><span class="pre">start()</span></code></a> before using any of Osprey’s other functions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>heapSizeMiB</strong> (<em>int</em>) – Size of the JVM heap in megabytes. This is essentially the amount of memory
Osprey will have to do computations. 1024 MiB is 1 GiB, but for larger designs,
you may want to use 2048 MiB (2 GiB), 4096 MiB (4 GiB), or even more memory.</li>
<li><strong>enableAssertions</strong> (<em>bool</em>) – pass <code class="docutils literal"><span class="pre">True</span></code> to enable JVM assertions. Only useful for debugging.</li>
<li><strong>stackSizeMiB</strong> (<em>int</em>) – Size of the JVM stack portion of the heap in megabytes.
Generally leave this at the default value, unless Osprey crashes because it’s too small. Then try increasing it.</li>
<li><strong>garbageSizeMiB</strong> (<em>int</em>) – Size of the garbage portion of the JVM heap that is reserved for temporary objects.
This default value is appropriate for the default heap size, but if using larger heap sizes, then increasing
the garbage size to 256, 512, or even 1024 MiB can give a modest improvement in performance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<p>Then you can call any of Osprey’s functions, described below.</p>
<hr class="docutils" />
<span class="target" id="module-osprey"></span><dl class="data">
<dt id="osprey.WILD_TYPE">
<code class="descclassname">osprey.</code><code class="descname">WILD_TYPE</code><em class="property"> = &quot;__WT__&quot;</em><a class="headerlink" href="#osprey.WILD_TYPE" title="Permalink to this definition">¶</a></dt>
<dd><p>Magic constant that refers to the wild-type template of the residue</p>
</dd></dl>

<dl class="function">
<dt id="osprey.readTextFile">
<code class="descclassname">osprey.</code><code class="descname">readTextFile</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.readTextFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Useful for passing file contents to functions that expect strings rather than file paths</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the text of the file at <code class="docutils literal"><span class="pre">path</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.readPdb">
<code class="descclassname">osprey.</code><code class="descname">readPdb</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.readPdb" title="Permalink to this definition">¶</a></dt>
<dd><p>loads a PDB file into a molecule object</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Raw molecules cannot be used directly in designs.
Create a <a class="reference internal" href="#osprey.Strand" title="osprey.Strand"><code class="xref py py-meth docutils literal"><span class="pre">Strand()</span></code></a> with the molecule to use it in a design.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<em>str</em>) – path to PDB file</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="api.structure.Molecule.html" title="edu.duke.cs.osprey.structure.Molecule"><code class="xref java java-ref docutils literal"><span class="pre">Molecule</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.writePdb">
<code class="descclassname">osprey.</code><code class="descname">writePdb</code><span class="sig-paren">(</span><em>mol</em>, <em>path</em>, <em>comment=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.writePdb" title="Permalink to this definition">¶</a></dt>
<dd><p>save a molecule to a PDB file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mol</strong> (<a class="reference internal" href="api.structure.Molecule.html" title="edu.duke.cs.osprey.structure.Molecule"><code class="xref java java-ref docutils literal"><span class="pre">Molecule</span></code></a>) – the molecule to save</li>
<li><strong>path</strong> (<em>str</em>) – path of the PDB file</li>
<li><strong>comment</strong> (<em>str</em>) – Optional comment to add to the PDB file headers</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.printGpuInfo">
<code class="descclassname">osprey.</code><code class="descname">printGpuInfo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#osprey.printGpuInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints information about GPU hardware present in the system
and which GPUs are usable by Osprey.</p>
<p>Both Cuda and OpenCL APIs are queried.</p>
</dd></dl>

<dl class="function">
<dt id="osprey.initExternalMemory">
<code class="descclassname">osprey.</code><code class="descname">initExternalMemory</code><span class="sig-paren">(</span><em>internalSizeMiB</em>, <em>tempDir=&lt;system temp dir&gt;</em>, <em>tempSubdir=&lt;automatically generated&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.initExternalMemory" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes external memory for calculations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>internalSizeMiB</strong> (<em>int</em>) – Set the maximum amount of internal memory (eg, RAM) to use for
large data structures. External memory-aware data structures will
use external memory (eg, disk, SSD, NAS) for extra storage space
when internal memory limits have been reached.</li>
<li><strong>tempDir</strong> (<em>str</em>) – Path to temporary directory to host external memory</li>
<li><strong>tempSubdir</strong> (<em>str</em>) – name of subdirectory within tempDir</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.Parallelism">
<code class="descclassname">osprey.</code><code class="descname">Parallelism</code><span class="sig-paren">(</span><em>cpuCores=1</em>, <em>gpus=0</em>, <em>streamsPerGpu=1</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.Parallelism" title="Permalink to this definition">¶</a></dt>
<dd><p>Specified how Osprey should use available hardware.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>cpuCores</strong> (<em>int</em>) – The number of CPUs cores to use.</li>
<li><strong>gpus</strong> (<em>int</em>) – The number of GPUs to use.</li>
<li><strong>streamsPerGpu</strong> (<em>int</em>) – The number of simultaneous tasks that should be given to each GPU</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.parallelism.Parallelism.html" title="edu.duke.cs.osprey.parallelism.Parallelism"><code class="xref java java-ref docutils literal"><span class="pre">Parallelism</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.TemplateLibrary">
<code class="descclassname">osprey.</code><code class="descname">TemplateLibrary</code><span class="sig-paren">(</span><em>forcefield=ForcefieldParams.Forcefield.AMBER</em>, <em>defaultTemplates=True</em>, <em>extraTemplates=[]</em>, <em>defaultTemplateCoords=True</em>, <em>extraTemplateCoords=[]</em>, <em>defaultRotamers=True</em>, <em>extraRotamers=[]</em>, <em>extraBackboneDependentRotamers=[]</em>, <em>defaultResidueEntropies=True</em>, <em>extraResidueEntropies=[]</em>, <em>makeDAminoAcids=True</em>, <em>moleculesForWildTypeRotamers=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.TemplateLibrary" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Library of <a class="reference internal" href="api.restypes.ResidueTemplate.html" title="edu.duke.cs.osprey.restypes.ResidueTemplate"><code class="xref java java-ref docutils literal"><span class="pre">ResidueTemplate</span></code></a> instances used by <a class="reference internal" href="api.confspace.Strand.html" title="edu.duke.cs.osprey.confspace.Strand"><code class="xref java java-ref docutils literal"><span class="pre">Strand</span></code></a> instances to define discrete flexibilty.</p>
<p>See Python script examples/python.GMEC/templateLibrary.py in your Osprey distribution for example usage.</p>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>forcefield</strong> (<a class="reference internal" href="api.energy.forcefield.ForcefieldParams.html" title="edu.duke.cs.osprey.energy.forcefield.ForcefieldParams"><code class="xref java java-ref docutils literal"><span class="pre">ForcefieldParams</span></code></a>) – The templates contain forcefield information like atom types,
so they need to be matched with a specific forcefield.</li>
<li><strong>defaultTemplates</strong> (<em>bool</em>) – By default, templates for natual amino acids are added</li>
<li><strong>extraTemplates</strong> (<em>[</em><em>template string</em><em> or </em><em>file path</em><em>]</em>) – Templates to add to the library</li>
<li><strong>defaultTemplateCoords</strong> (<em>bool</em>) – By default, template coordinates for natual amino acids are added</li>
<li><strong>extraTemplateCoords</strong> (<em>[</em><em>coords string</em><em> or </em><em>file path</em><em>]</em>) – <p>Template coordinates to add to the library</p>
<span class="target" id="index-0"></span><div class="admonition-todo admonition">
<p class="first admonition-title">Todo</p>
<p class="last">explain template coord file format. anyone know about this?</p>
</div>
</li>
<li><strong>defaultRotamers</strong> (<em>bool</em>) – <p>By default, the Lovell rotamer library is added.</p>
<p>See <a class="reference internal" href="api.restypes.ResidueTemplateLibrary.html#ResidueTemplateLibrary.addLovellRotamers" title="edu.duke.cs.osprey.restypes.ResidueTemplateLibrary#addLovellRotamers"><code class="xref java java-ref docutils literal"><span class="pre">addLovellRotamers</span></code></a></p>
</li>
<li><strong>extraRotamers</strong> (<em>[</em><em>rotamers string</em><em> or </em><em>file path</em><em>]</em>) – <p>Rotamers to add to the library</p>
<span class="target" id="index-1"></span><div class="admonition-todo admonition">
<p class="first admonition-title">Todo</p>
<p class="last">explain rotamer file format. anyone know about this?</p>
</div>
</li>
<li><strong>extraBackboneDependentRotamers</strong> (<em>[</em><em>backbone-dependent rotamers string</em><em> or </em><em>file path</em><em>]</em>) – <p>Backbone-dependent rotamers to add to the library</p>
<span class="target" id="index-2"></span><div class="admonition-todo admonition">
<p class="first admonition-title">Todo</p>
<p class="last">explain backbone-dependent rotamer file format. anyone know about this?</p>
</div>
</li>
<li><strong>makeDAminoAcids</strong> (<em>bool</em>) – True to generate D amino acids for the template library</li>
<li><strong>moleculesForWildTypeRotamers</strong> ([<a class="reference internal" href="api.structure.Molecule.html" title="edu.duke.cs.osprey.structure.Molecule"><code class="xref java java-ref docutils literal"><span class="pre">Molecule</span></code></a>]) – Add a molecule to include its wild-type rotamers in the library</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.restypes.ResidueTemplateLibrary.html" title="edu.duke.cs.osprey.restypes.ResidueTemplateLibrary"><code class="xref java java-ref docutils literal"><span class="pre">ResidueTemplateLibrary</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.Strand">
<code class="descclassname">osprey.</code><code class="descname">Strand</code><span class="sig-paren">(</span><em>pathOrMol</em>, <em>residues=None</em>, <em>templateLib=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.Strand" title="Permalink to this definition">¶</a></dt>
<dd><p>A molecule with associated residue flexibility information.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pathOrMol</strong> (str or <a class="reference internal" href="api.structure.Molecule.html" title="edu.duke.cs.osprey.structure.Molecule"><code class="xref java java-ref docutils literal"><span class="pre">Molecule</span></code></a>) – path to a PDB file, or a molecule instance</li>
<li><strong>residues</strong> (<em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – range of residue numbers, inclusive. <cite>None</cite> to include all residues.</li>
<li><strong>templateLib</strong> (<a class="reference internal" href="api.restypes.ResidueTemplateLibrary.html" title="edu.duke.cs.osprey.restypes.ResidueTemplateLibrary"><code class="xref java java-ref docutils literal"><span class="pre">ResidueTemplateLibrary</span></code></a>) – <p>The template library to use for this strand.</p>
<p>If no template library is specified, a default template library is created.</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.confspace.Strand.html" title="edu.duke.cs.osprey.confspace.Strand"><code class="xref java java-ref docutils literal"><span class="pre">Strand</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.ConfSpace">
<code class="descclassname">osprey.</code><code class="descname">ConfSpace</code><span class="sig-paren">(</span><em>strands</em>, <em>shellDist=float('inf')</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.ConfSpace" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Maintains the design positions and residue conformations for a list of strands.</p>
<p>Also creates molecules in conformations on-demand.</p>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>strands</strong> (<a class="reference internal" href="api.confspace.Strand.html" title="edu.duke.cs.osprey.confspace.Strand"><code class="xref java java-ref docutils literal"><span class="pre">Strand</span></code></a> or list of Strands) – the strands to use</li>
<li><strong>shellDist</strong> (<em>float</em>) – A residue is included in the steric shell if any of its atoms lies within
shellDist angstroms of any atom in any flexible residue.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.ForcefieldParams">
<code class="descclassname">osprey.</code><code class="descname">ForcefieldParams</code><span class="sig-paren">(</span><em>forcefield=Forcefield.AMBER</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.ForcefieldParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Options for configuring forcefields for energy calculation.</p>
<p>Configure the forcefield parameters by setting the properties of the <a class="reference internal" href="api.energy.forcefield.ForcefieldParams.html" title="edu.duke.cs.osprey.energy.forcefield.ForcefieldParams"><code class="xref java java-ref docutils literal"><span class="pre">ForcefieldParams</span></code></a> object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>forcefield</strong> (<a class="reference internal" href="api.energy.forcefield.ForcefieldParams.Forcefield.html" title="edu.duke.cs.osprey.energy.forcefield.ForcefieldParams$Forcefield"><code class="xref java java-ref docutils literal"><span class="pre">Forcefield</span></code></a>) – </td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="api.energy.forcefield.ForcefieldParams.html" title="edu.duke.cs.osprey.energy.forcefield.ForcefieldParams"><code class="xref java java-ref docutils literal"><span class="pre">ForcefieldParams</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.EnergyCalculator">
<code class="descclassname">osprey.</code><code class="descname">EnergyCalculator</code><span class="sig-paren">(</span><em>confSpace</em>, <em>ffparams</em>, <em>parallelism=Parallelism.makeCpu(1)</em>, <em>type=None</em>, <em>isMinimizing=True</em>, <em>infiniteWellEnergy=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.EnergyCalculator" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Computes the energy of a molecule fragment from a conformation space using the desired forcefield parameters
and residue interactions.</p>
<p>Residue interactions are specified via <a class="reference internal" href="api.energy.ResidueInteractions.html" title="edu.duke.cs.osprey.energy.ResidueInteractions"><code class="xref java java-ref docutils literal"><span class="pre">ResidueInteractions</span></code></a> instances.
Forcefield implementations are chosen via the type argument. If no type is specified, the best forcefield
available implementation is automatically chosen based on the parallelism argument.</p>
<p>If a fragment has continuous degrees of freedom, minimization will be performed before forcefield evaluation.</p>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>confSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – The conformation space containing the residue templates to use for atom connectivities.</li>
<li><strong>ffparams</strong> (<a class="reference internal" href="api.energy.forcefield.ForcefieldParams.html" title="edu.duke.cs.osprey.energy.forcefield.ForcefieldParams"><code class="xref java java-ref docutils literal"><span class="pre">ForcefieldParams</span></code></a>) – </li>
<li><strong>parallelism</strong> (<a class="reference internal" href="api.parallelism.Parallelism.html" title="edu.duke.cs.osprey.parallelism.Parallelism"><code class="xref java java-ref docutils literal"><span class="pre">Parallelism</span></code></a>) – The parallel hardware available for Osprey calculations.
CPU thread pools are available, as well as high-performance GPU kernels.
Just describe your available hardware, and Osprey will try to pick the best implementation.
If you need more control, try the setType() option directly.</li>
<li><strong>type</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.Type.html" title="edu.duke.cs.osprey.energy.EnergyCalculator$Type"><code class="xref java java-ref docutils literal"><span class="pre">Type</span></code></a>) – <p>Directly choose the energy calculator implementation.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Not all energy calculator implementations may be compatible with your design’s conformation space. Choosing an incompatible implementation can cause Osprey to crash.  If you’re unsure which implementation to use, just use setParallelism() instead of setting the type directly.</p>
</div>
</li>
<li><strong>isMinimizing</strong> (<em>bool</em>) – True to minimize continuous degrees of freedom in conformations. False to use only rigid structures.</li>
<li><strong>infiniteWellEnergy</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – <p>When set, if a minimized energy is below this threshold, assume the conformation
fell into an infinite well due to a severe clash (this is a known flaw of the AMBER
forcefield, all parameterizations). Then attempt the minimization again,
but try to resolve the clash first with a purely vdW forcefield. If the clash cannot
be resolved, and the conformation falls into the well again, return positive infinity
energy instead of effectively negative infinity.</p>
<p>If you wish to use the inifinite well detection and resolution, -10,000 kcal/mol is
probably a good threshold to choose.</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.SharedEnergyCalculator">
<code class="descclassname">osprey.</code><code class="descname">SharedEnergyCalculator</code><span class="sig-paren">(</span><em>ecalc</em>, <em>isMinimizing=True</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.SharedEnergyCalculator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an energy calculator that shares resources with an existing energy calculator,
(including thread pools, GPUs, atom connectivity cache, forcefield),
but can have different configuration.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ecalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – The existing energy calculator with which to share resources</li>
<li><strong>isMinimizing</strong> (<em>bool</em>) – True to minimize continuous degrees of freedom in conformations. False to use only rigid structures.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.ConfEnergyCalculator">
<code class="descclassname">osprey.</code><code class="descname">ConfEnergyCalculator</code><span class="sig-paren">(</span><em>confSpace</em>, <em>ecalc</em>, <em>referenceEnergies=None</em>, <em>addResEntropy=False</em>, <em>energyPartition=EnergyPartition.Traditional</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.ConfEnergyCalculator" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Calculate energy for molecules created from conformation spaces.</p>
<p>Provides support for applying conformation energy modifications,
such as reference energies, residue entropies, and energy partitions.</p>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>confSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – </li>
<li><strong>ecalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – </li>
<li><strong>referenceEnergies</strong> (<a class="reference internal" href="api.ematrix.SimpleReferenceEnergies.html" title="edu.duke.cs.osprey.ematrix.SimpleReferenceEnergies"><code class="xref java java-ref docutils literal"><span class="pre">SimpleReferenceEnergies</span></code></a>) – </li>
<li><strong>addResEntropy</strong> (<em>bool</em>) – </li>
<li><strong>energyPartition</strong> (<a class="reference internal" href="api.energy.EnergyPartition.html" title="edu.duke.cs.osprey.energy.EnergyPartition"><code class="xref java java-ref docutils literal"><span class="pre">EnergyPartition</span></code></a>) – How energies should be partitioned among single and pair fragments.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.energy.ConfEnergyCalculator.html" title="edu.duke.cs.osprey.energy.ConfEnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculator</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.EnergyMatrix">
<code class="descclassname">osprey.</code><code class="descname">EnergyMatrix</code><span class="sig-paren">(</span><em>confEcalc</em>, <em>cacheFile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.EnergyMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a matrix of energies between pairs of residue conformations to be used by A* search.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>confEcalc</strong> (<a class="reference internal" href="api.energy.ConfEnergyCalculator.html" title="edu.duke.cs.osprey.energy.ConfEnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculator</span></code></a>) – How conformation energies should be calculated.</li>
<li><strong>cacheFile</strong> (<code class="docutils literal"><code class="xref java java-ref docutils literal"><span class="pre">File</span></code></code>) – <p>Path to file where energy matrix should be saved between computations.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Energy matrix computation can take a long time, but often the results can be reused between computations. Use a cache file to skip energy matrix computation on the next Osprey run if the energy matrix has already been computed once before.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">If design settings are changed between runs, Osprey will make some effort to detect that the energy matrix cache is out-of-date and compute a new energy matrix instead of usng the cached, incorrect one. Osprey might not detect all design changes though, and incorrectly reuse a cached energy matrix, so it is best to manually delete the entry matrix cache file after changing design settings.</p>
</div>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.ReferenceEnergies">
<code class="descclassname">osprey.</code><code class="descname">ReferenceEnergies</code><span class="sig-paren">(</span><em>confSpace</em>, <em>ecalc</em>, <em>addResEntropy=False</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.ReferenceEnergies" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates a reference energy for each residue position and residue type
based on the minimum energy of all residue conformations at that position
and residue type.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>confSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – </li>
<li><strong>ecalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – </li>
<li><strong>addResEntropy</strong> (<em>bool</em>) – </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.ematrix.SimpleReferenceEnergies.html" title="edu.duke.cs.osprey.ematrix.SimpleReferenceEnergies"><code class="xref java java-ref docutils literal"><span class="pre">SimpleReferenceEnergies</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.DEE">
<code class="descclassname">osprey.</code><code class="descname">DEE</code><span class="sig-paren">(</span><em>confSpace</em>, <em>emat</em>, <em>singlesThreshold=100.0</em>, <em>pairsThreshold=100.0</em>, <em>singlesGoldsteinDiffThreshold=None</em>, <em>pairsGoldsteinDiffThreshold=None</em>, <em>typeDependent=False</em>, <em>numIterations=Integer.MAX_VALUE</em>, <em>showProgress=False</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.DEE" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Dead-End Elimination (DEE) to remove tuples of RCs that will not appear in low-energy conformations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>confSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – The design conformation space</li>
<li><strong>emat</strong> (<a class="reference internal" href="api.ematrix.EnergyMatrix.html" title="edu.duke.cs.osprey.ematrix.EnergyMatrix"><code class="xref java java-ref docutils literal"><span class="pre">EnergyMatrix</span></code></a>) – An energy matrix computed for the conformation space</li>
<li><strong>singlesThreshold</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – </li>
<li><strong>pairsThreshold</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – </li>
<li><strong>singlesGoldsteinDiffThreshold</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – </li>
<li><strong>pairsGoldsteinDiffThreshold</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – </li>
<li><strong>typeDependent</strong> (<em>bool</em>) – </li>
<li><strong>numIterations</strong> (<em>int</em>) – </li>
<li><strong>showProgress</strong> (<em>bool</em>) – </li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.AStarTraditional">
<code class="descclassname">osprey.</code><code class="descname">AStarTraditional</code><span class="sig-paren">(</span><em>emat</em>, <em>confSpaceOrPmat</em>, <em>showProgress=True</em>, <em>useExternalMemory=False</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.AStarTraditional" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Uses the traditional estimation function to guide the tree search.
<a class="reference internal" href="#leach1998" id="id1">[Leach1998]</a></p>
<table class="docutils citation" frame="void" id="leach1998" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Leach1998]</td><td>Leach, A.R. and Lemon, A.P., 1998. Exploring the conformational space of protein side chains using dead-end elimination and the A* algorithm. Proteins Structure Function and Genetics, 33(2), pp.227-239.</td></tr>
</tbody>
</table>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>emat</strong> (<a class="reference internal" href="api.ematrix.EnergyMatrix.html" title="edu.duke.cs.osprey.ematrix.EnergyMatrix"><code class="xref java java-ref docutils literal"><span class="pre">EnergyMatrix</span></code></a>) – The energy matrix to use for pairwise residue conformation energies.</li>
<li><strong>confSpaceOrPmat</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a> or <a class="reference internal" href="api.pruning.PruningMatrix.html" title="edu.duke.cs.osprey.pruning.PruningMatrix"><code class="xref java java-ref docutils literal"><span class="pre">PruningMatrix</span></code></a>) – The conformation space containing the residue conformations to search.</li>
<li><strong>useExternalMemory</strong> (<em>boolean</em>) – <p>set to True to use external memory.</p>
<p>Use external memory (eg, disk, SSD, NAS) when large A* searches
cannot fit in internal memory (eg, RAM).<p>Use <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setInternalLimit" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setInternalLimit"><code class="xref java java-ref docutils literal"><span class="pre">setInternalLimit</span></code></a> to set the amount of fixed internal memory
and <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setTempDir" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setTempDir"><code class="xref java java-ref docutils literal"><span class="pre">setTempDir</span></code></a> to set the file path for external memory.</p>
</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.astar.conf.ConfAStarTree.html" title="edu.duke.cs.osprey.astar.conf.ConfAStarTree"><code class="xref java java-ref docutils literal"><span class="pre">ConfAStarTree</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.AStarMPLP">
<code class="descclassname">osprey.</code><code class="descname">AStarMPLP</code><span class="sig-paren">(</span><em>emat</em>, <em>confSpaceOrPmat</em>, <em>updater=NodeUpdater()</em>, <em>numIterations=1</em>, <em>convergenceThreshold=0.0001</em>, <em>useExternalMemory=False</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.AStarMPLP" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Creates an A* search using a newer estimation function based on Max Product Linear
Programming (MPLP).
<a class="reference internal" href="#globerson2008" id="id1">[Globerson2008]</a></p>
<p>For large designs, this A* implementation can be dramatically faster than the traditional
one, and often require much less memory too.</p>
<table class="docutils citation" frame="void" id="globerson2008" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Globerson2008]</td><td>Globerson, A. and Jaakkola, T.S., 2008. Fixing max-product: Convergent message passing algorithms for MAP LP-relaxations. In Advances in neural information processing systems (pp. 553-560).</td></tr>
</tbody>
</table>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>emat</strong> (<a class="reference internal" href="api.ematrix.EnergyMatrix.html" title="edu.duke.cs.osprey.ematrix.EnergyMatrix"><code class="xref java java-ref docutils literal"><span class="pre">EnergyMatrix</span></code></a>) – The energy matrix to use for pairwise residue conformation energies.</li>
<li><strong>confSpaceOrPmat</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a> or <a class="reference internal" href="api.pruning.PruningMatrix.html" title="edu.duke.cs.osprey.pruning.PruningMatrix"><code class="xref java java-ref docutils literal"><span class="pre">PruningMatrix</span></code></a>) – The conformation space containing the residue conformations to search.</li>
<li><strong>updater</strong> (<a class="reference internal" href="api.astar.conf.scoring.mplp.MPLPUpdater.html" title="edu.duke.cs.osprey.astar.conf.scoring.mplp.MPLPUpdater"><code class="xref java java-ref docutils literal"><span class="pre">MPLPUpdater</span></code></a>) – <p>The type of MPLP update step to use for each iteration.</p>
<p>There are two options: <a class="reference internal" href="api.astar.conf.scoring.mplp.EdgeUpdater.html" title="edu.duke.cs.osprey.astar.conf.scoring.mplp.EdgeUpdater"><code class="xref java java-ref docutils literal"><span class="pre">EdgeUpdater</span></code></a> and <a class="reference internal" href="api.astar.conf.scoring.mplp.NodeUpdater.html" title="edu.duke.cs.osprey.astar.conf.scoring.mplp.NodeUpdater"><code class="xref java java-ref docutils literal"><span class="pre">NodeUpdater</span></code></a>.
In practice, the EdgeUpdater seems to work best when reference energies
are used. When reference energies are not used, the NodeUpdater seems
to work best.</p>
</li>
<li><strong>numIterations</strong> (<em>int</em>) – <p>The number of MPLP iterations to execute on each A* node.</p>
<p>This value doesn’t affect the accuracy of the conformation search, only the speed.</p>
<p>The more iterations, the more accurate the A* estimation function will be,
and fewer nodes will need to be explored to reach a leaf node. The tradeoff though is
increased compute time per node explored.</p>
<p>Generally, it’s safe to start with one iteration, then experimentally try more
iterations to see if it reduces the total A* search time.</p>
</li>
<li><strong>convergenceThreshold</strong> (<em>float</em>) – <p>If the change in energy after an iteration of the estimation function is below this
threshold, MPLP will stop iterating.</p>
<p>This value doesn’t affect the accuracy of the conformation search, only the speed.</p>
<p>It also has no effect if the number of iterations is 1.</p>
<p>For a larger number of iterations, increasing this value may reduce the time spent
at each node, at the cost of exploring more total nodes. Decreasing this value may
increase time spent at each node, but not necessarily reduce the total number of
nodes explored.</p>
<p>Generally, this value won’t need to be adjusted for most designs. For designs with
large numbers of MPLP iterations, optimizing this value may increase performance though.</p>
</li>
<li><strong>useExternalMemory</strong> (<em>boolean</em>) – <p>set to True to use external memory.</p>
<p>Use external memory (eg, disk, SSD, NAS) when large A* searches
cannot fit in internal memory (eg, RAM).<p>Use <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setInternalLimit" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setInternalLimit"><code class="xref java java-ref docutils literal"><span class="pre">setInternalLimit</span></code></a> to set the amount of fixed internal memory
and <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setTempDir" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setTempDir"><code class="xref java java-ref docutils literal"><span class="pre">setTempDir</span></code></a> to set the file path for external memory.</p>
</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.astar.conf.ConfAStarTree.html" title="edu.duke.cs.osprey.astar.conf.ConfAStarTree"><code class="xref java java-ref docutils literal"><span class="pre">ConfAStarTree</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.GMECFinder">
<code class="descclassname">osprey.</code><code class="descname">GMECFinder</code><span class="sig-paren">(</span><em>astar</em>, <em>confEcalc</em>, <em>confLog=None</em>, <em>printIntermediateConfs=False</em>, <em>useExternalMemory=False</em>, <em>resumeLog=None</em>, <em>confDBFile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.GMECFinder" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Searches a conformation space for the single conformation that minimizes the
desired energy function, ie the Global Minimum Energy Conformation, or GMEC.</p>
<p>The search is performed by ordering all conformations in the conformation space
using A* search, and evaluating the energy of each conformation in order.</p>
<p>The search terminates returns the lowest-energy conformation found so far
when it can be proven that all remaining conformations in the conformation
space must have higher energy.</p>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>astar</strong> (<a class="reference internal" href="api.confspace.ConfSearch.html" title="edu.duke.cs.osprey.confspace.ConfSearch"><code class="xref java java-ref docutils literal"><span class="pre">ConfSearch</span></code></a>) – <p>A* implementation to sort conformations in the conformation space.</p>
<p>Use one of <a class="reference internal" href="#osprey.AStarTraditional" title="osprey.AStarTraditional"><code class="xref py py-func docutils literal"><span class="pre">AStarTraditional()</span></code></a> or <a class="reference internal" href="#osprey.AStarMPLP" title="osprey.AStarMPLP"><code class="xref py py-func docutils literal"><span class="pre">AStarMPLP()</span></code></a> to get an A* implementation.</p>
</li>
<li><strong>confEcalc</strong> (<a class="reference internal" href="api.energy.ConfEnergyCalculator.html" title="edu.duke.cs.osprey.energy.ConfEnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculator</span></code></a>) – <p>Calculates the energy for a conformation.</p>
<p>Use <a class="reference internal" href="#osprey.ConfEnergyCalculator" title="osprey.ConfEnergyCalculator"><code class="xref py py-func docutils literal"><span class="pre">ConfEnergyCalculator()</span></code></a> to get a conformation energy calculator.</p>
</li>
<li><strong>confLog</strong> (<em>str</em>) – Path to file where conformations found during conformation space search should be logged.</li>
<li><strong>printIntermediateConfs</strong> (<em>bool</em>) – True to print all conformations found during A* search to the console</li>
<li><strong>useExternalMemory</strong> (<em>bool</em>) – <p>True to use external memory (eg, disk, SSD, NAS) for when large data
structures cannot fit in internal memory (eg, RAM).</p>
<p>Use <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setInternalLimit" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setInternalLimit"><code class="xref java java-ref docutils literal"><span class="pre">setInternalLimit</span></code></a> to set the amount of fixed internal memory
and <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setTempDir" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setTempDir"><code class="xref java java-ref docutils literal"><span class="pre">setTempDir</span></code></a> to set the file path for external memory.</p>
</li>
<li><strong>resumeLog</strong> (<em>str</em>) – Path to log file where resume info will be written or read, so designs can be resumed.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.gmec.SimpleGMECFinder.html" title="edu.duke.cs.osprey.gmec.SimpleGMECFinder"><code class="xref java java-ref docutils literal"><span class="pre">SimpleGMECFinder</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.DEEGMECFinder">
<code class="descclassname">osprey.</code><code class="descname">DEEGMECFinder</code><span class="sig-paren">(</span><em>emat</em>, <em>confSpace</em>, <em>ecalc</em>, <em>confEcalc</em>, <em>name</em>, <em>use_epic</em>, <em>use_lute</em>, <em>confLog=None</em>, <em>printIntermediateConfs=False</em>, <em>useExternalMemory=False</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.DEEGMECFinder" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Searches a conformation space for the single conformation that minimizes the
desired energy function, ie the Global Minimum Energy Conformation, or GMEC.</p>
<p>The search is performed by ordering all conformations in the conformation space
using A* search, and evaluating the energy of each conformation in order.</p>
<p>The search terminates returns the lowest-energy conformation found so far
when it can be proven that all remaining conformations in the conformation
space must have higher energy.</p>
</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>astar</strong> (<a class="reference internal" href="api.confspace.ConfSearch.html" title="edu.duke.cs.osprey.confspace.ConfSearch"><code class="xref java java-ref docutils literal"><span class="pre">ConfSearch</span></code></a>) – <p>A* implementation to sort conformations in the conformation space.</p>
<p>Use one of <a class="reference internal" href="#osprey.AStarTraditional" title="osprey.AStarTraditional"><code class="xref py py-func docutils literal"><span class="pre">AStarTraditional()</span></code></a> or <a class="reference internal" href="#osprey.AStarMPLP" title="osprey.AStarMPLP"><code class="xref py py-func docutils literal"><span class="pre">AStarMPLP()</span></code></a> to get an A* implementation.</p>
</li>
<li><strong>confEcalc</strong> (<a class="reference internal" href="api.energy.ConfEnergyCalculator.html" title="edu.duke.cs.osprey.energy.ConfEnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculator</span></code></a>) – <p>Calculates the energy for a conformation.</p>
<p>Use <a class="reference internal" href="#osprey.ConfEnergyCalculator" title="osprey.ConfEnergyCalculator"><code class="xref py py-func docutils literal"><span class="pre">ConfEnergyCalculator()</span></code></a> to get a conformation energy calculator.</p>
</li>
<li><strong>confLog</strong> (<em>str</em>) – Path to file where conformations found during conformation space search should be logged.</li>
<li><strong>printIntermediateConfs</strong> (<em>bool</em>) – True to print all conformations found during A* search to the console</li>
<li><strong>useExternalMemory</strong> (<em>bool</em>) – <p>True to use external memory (eg, disk, SSD, NAS) for when large data
structures cannot fit in internal memory (eg, RAM).</p>
<p>Use <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setInternalLimit" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setInternalLimit"><code class="xref java java-ref docutils literal"><span class="pre">setInternalLimit</span></code></a> to set the amount of fixed internal memory
and <a class="reference internal" href="api.externalMemory.ExternalMemory.html#ExternalMemory.setTempDir" title="edu.duke.cs.osprey.externalMemory.ExternalMemory#setTempDir"><code class="xref java java-ref docutils literal"><span class="pre">setTempDir</span></code></a> to set the file path for external memory.</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.gmec.SimpleGMECFinder.html" title="edu.duke.cs.osprey.gmec.SimpleGMECFinder"><code class="xref java java-ref docutils literal"><span class="pre">SimpleGMECFinder</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.KStar">
<code class="descclassname">osprey.</code><code class="descname">KStar</code><span class="sig-paren">(</span><em>proteinConfSpace</em>, <em>ligandConfSpace</em>, <em>complexConfSpace</em>, <em>ecalc</em>, <em>confEcalcFactory</em>, <em>astarFactory</em>, <em>epsilon=0.683</em>, <em>stabilityThreshold=5.0</em>, <em>maxSimultaneousMutations=1</em>, <em>energyMatrixCachePattern=None</em>, <em>confDBPattern=None</em>, <em>writeSequencesToConsole=False</em>, <em>writeSequencesToFile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.KStar" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Implementation of the K* algorithm to predict protein sequence mutations that improve
binding affinity by computing provably accurate Boltzmann-weighted ensembles
<a class="reference internal" href="#lilien2008" id="id1">[Lilien2008]</a>.</p>
<table class="docutils citation" frame="void" id="lilien2008" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Lilien2008]</td><td>Ryan H. Lilien, Brian W. Stevens, Amy C. Anderson, and Bruce R. Donald, 2005. A Novel Ensemble-Based Scoring and Search Algorithm for Protein Redesign and Its Application to Modify the Substrate Specificity of the Gramicidin Synthetase A Phenylalanine Adenylation Enzyme In Journal of Computational Biology (vol 12. num. 6 pp. 740–761).</td></tr>
</tbody>
</table>
</p>
<p>For examples using K*, see the examples/python.KStar directory in your Osprey distribution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>proteinConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing just the protein strand</li>
<li><strong>ligandConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing just the ligand strand</li>
<li><strong>complexConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing both the protein and ligand strands</li>
<li><strong>ecalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – Calculates the energy for a molecule</li>
<li><strong>confEcalcFactory</strong> (<a class="reference internal" href="api.kstar.KStar.ConfEnergyCalculatorFactory.html" title="edu.duke.cs.osprey.kstar.KStar$ConfEnergyCalculatorFactory"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculatorFactory</span></code></a>) – A function that makes a ConfEnergyCalculator with the desired options</li>
<li><strong>astarFactory</strong> (<a class="reference internal" href="api.kstar.KStar.ConfSearchFactory.html" title="edu.duke.cs.osprey.kstar.KStar$ConfSearchFactory"><code class="xref java java-ref docutils literal"><span class="pre">ConfSearchFactory</span></code></a>) – A function that makes a ConfSearchFactory (e.g, A* search) with the desired options</li>
<li><strong>epsilon</strong> (<em>float</em>) – <p>Value of epsilon in (0,1] for the epsilon-approximation to a partition function.</p>
<p>Smaller values for epsilon yield more accurate predictions, but can take
longer to run.</p>
</li>
<li><strong>stabilityThreshold</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – <p>Pruning criteria to remove sequences with unstable unbound states relative to the wild type sequence.
Defined in units of kcal/mol.</p>
<p>More precisely, a sequence is pruned when the following expression is true:</p>
<p>U(Z_s) &lt; L(W_s) * B(t)</p>
<dl class="docutils">
<dt>where:</dt>
<dd><ul class="first last">
<li>s represents the unbound protein strand, or unbound ligand strand</li>
<li>U(Z_s) is the upper bound on the partition function for strand s</li>
<li>L(W_s) is the lower bound on the partition function for strand s in the wild type</li>
<li>t is the stability threshold</li>
<li>B() is the Boltzmann weighting function</li>
</ul>
</dd>
</dl>
<p>Set to null to disable the filter entirely.</p>
</li>
<li><strong>maxSimultaneousMutations</strong> (<em>int</em>) – The maximum number of simultaneous residue mutations to consider for each sequence mutant</li>
<li><strong>energyMatrixCachePattern</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – <p>Pattern of the filename to cache energy matrices.</p>
<p>K*-type algorithms must calculate multiple energy matrices.
By default, these energy matrices are not cached between runs.
To cache energy matrices between runs, supply a pattern such as:</p>
<p>”theFolder/emat.*.dat”</p>
<p>The * in the pattern is a wildcard character that will be replaced with
each type of energy matrix used by the K*-type algorithm.</p>
</li>
<li><strong>confDBPattern</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – <p>If a design experiences an unexpected abort, the conformation database can allow you to restore the
design state and resume the calculation close to where it was aborted.
Set a pattern to turn on the conf DB such as:</p>
<p>”theFolder/conf.*.db”</p>
<p>The * in the pattern is a wildcard character that will be replaced with
each type of energy matrix used by the K*-type algorithm.</p>
</li>
<li><strong>writeSequencesToConsole</strong> (<em>bool</em>) – True to write sequences and scores to the console</li>
<li><strong>writeSequencesToFile</strong> (<em>str</em>) – Path to the log file to write sequences scores (in TSV format), or None to skip logging</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.kstar.KStar.html" title="edu.duke.cs.osprey.kstar.KStar"><code class="xref java java-ref docutils literal"><span class="pre">KStar</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.BBKStar">
<code class="descclassname">osprey.</code><code class="descname">BBKStar</code><span class="sig-paren">(</span><em>proteinConfSpace</em>, <em>ligandConfSpace</em>, <em>complexConfSpace</em>, <em>rigidEcalc</em>, <em>minimizingEcalc</em>, <em>confEcalcFactory</em>, <em>astarFactory</em>, <em>epsilon=0.683</em>, <em>stabilityThreshold=5.0</em>, <em>maxSimultaneousMutations=1</em>, <em>energyMatrixCachePattern=None</em>, <em>confDBPattern=None</em>, <em>numBestSequences=1</em>, <em>numConfsPerBatch=8</em>, <em>writeSequencesToConsole=False</em>, <em>writeSequencesToFile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.BBKStar" title="Permalink to this definition">¶</a></dt>
<dd><p><p>Implementation of the BBK* algorithm to predict protein sequence mutations that improve
binding affinity by computing provably accurate Boltzmann-weighted ensembles
<a class="reference internal" href="#ojewole2008" id="id1">[Ojewole2008]</a>.</p>
<table class="docutils citation" frame="void" id="ojewole2008" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Ojewole2008]</td><td>Adegoke A. Ojewole, Jonathan D. Jou, Vance G. Fowler, Jr., and Bruce R. Donald, 2017. BBK* (Branch and Bound over K*): A Provable and Efficient Ensemble-Based Algorithm to Optimize Stability and Binding Affinity over Large Sequence Spaces In Research in Computational Molecular Biology (accepted, in press).</td></tr>
</tbody>
</table>
</p>
<p>For examples using BBK*, see the examples/python.KStar directory in your Osprey distribution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>proteinConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing just the protein strand</li>
<li><strong>ligandConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing just the ligand strand</li>
<li><strong>complexConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing both the protein and ligand strands</li>
<li><strong>rigidEcalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – Calculates the rigid-rotamer energy for a molecule</li>
<li><strong>minimizingEcalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – Calculates the continuous-rotamer (minimized) energy for a molecule</li>
<li><strong>confEcalcFactory</strong> (<a class="reference internal" href="api.kstar.KStar.ConfEnergyCalculatorFactory.html" title="edu.duke.cs.osprey.kstar.KStar$ConfEnergyCalculatorFactory"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculatorFactory</span></code></a>) – A function that makes a ConfEnergyCalculator with the desired options</li>
<li><strong>astarFactory</strong> (<a class="reference internal" href="api.kstar.KStar.ConfSearchFactory.html" title="edu.duke.cs.osprey.kstar.KStar$ConfSearchFactory"><code class="xref java java-ref docutils literal"><span class="pre">ConfSearchFactory</span></code></a>) – A function that makes a ConfSearchFactory (e.g, A* search) with the desired options</li>
<li><strong>epsilon</strong> (<em>float</em>) – <p>Value of epsilon in (0,1] for the epsilon-approximation to a partition function.</p>
<p>Smaller values for epsilon yield more accurate predictions, but can take
longer to run.</p>
</li>
<li><strong>stabilityThreshold</strong> (<code class="docutils literal"><span class="pre">Double</span></code>) – <p>Pruning criteria to remove sequences with unstable unbound states relative to the wild type sequence.
Defined in units of kcal/mol.</p>
<p>More precisely, a sequence is pruned when the following expression is true:</p>
<p>U(Z_s) &lt; L(W_s) * B(t)</p>
<dl class="docutils">
<dt>where:</dt>
<dd><ul class="first last">
<li>s represents the unbound protein strand, or unbound ligand strand</li>
<li>U(Z_s) is the upper bound on the partition function for strand s</li>
<li>L(W_s) is the lower bound on the partition function for strand s in the wild type</li>
<li>t is the stability threshold</li>
<li>B() is the Boltzmann weighting function</li>
</ul>
</dd>
</dl>
<p>Set to null to disable the filter entirely.</p>
</li>
<li><strong>maxSimultaneousMutations</strong> (<em>int</em>) – The maximum number of simultaneous residue mutations to consider for each sequence mutant</li>
<li><strong>energyMatrixCachePattern</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – <p>Pattern of the filename to cache energy matrices.</p>
<p>K*-type algorithms must calculate multiple energy matrices.
By default, these energy matrices are not cached between runs.
To cache energy matrices between runs, supply a pattern such as:</p>
<p>”theFolder/emat.*.dat”</p>
<p>The * in the pattern is a wildcard character that will be replaced with
each type of energy matrix used by the K*-type algorithm.</p>
</li>
<li><strong>confDBPattern</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – <p>If a design experiences an unexpected abort, the conformation database can allow you to restore the
design state and resume the calculation close to where it was aborted.
Set a pattern to turn on the conf DB such as:</p>
<p>”theFolder/conf.*.db”</p>
<p>The * in the pattern is a wildcard character that will be replaced with
each type of energy matrix used by the K*-type algorithm.</p>
</li>
<li><strong>numBestSequences</strong> (<em>int</em>) – The number of best (by K* score) sequences to evaluate before finishing</li>
<li><strong>numConfsPerBatch</strong> (<em>int</em>) – <p>The number of conformations to evaluate per batch of partition function refinement</p>
<p>For best results, make this a multiple of the available parallelism. e.g., if using 4 threads,
try a batch size of 4, 8, 12, 16, etc.</p>
</li>
<li><strong>writeSequencesToConsole</strong> (<em>bool</em>) – True to write sequences and scores to the console</li>
<li><strong>writeSequencesToFile</strong> (<em>str</em>) – Path to the log file to write sequences scores (in TSV format), or None to skip logging</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.kstar.BBKStar.html" title="edu.duke.cs.osprey.kstar.BBKStar"><code class="xref java java-ref docutils literal"><span class="pre">BBKStar</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.ConfAnalyzer">
<code class="descclassname">osprey.</code><code class="descname">ConfAnalyzer</code><span class="sig-paren">(</span><em>confEcalc</em>, <em>emat</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.ConfAnalyzer" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows information about a single conformation.</p>
<p>For examples using the conf analyzer, see examples/python.GMEC/analyzeConf.py in your Osprey distribution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>confEcalc</strong> (<a class="reference internal" href="api.energy.ConfEnergyCalculator.html" title="edu.duke.cs.osprey.energy.ConfEnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculator</span></code></a>) – Calculates the energy for a conformation.</li>
<li><strong>emat</strong> (<a class="reference internal" href="api.ematrix.EnergyMatrix.html" title="edu.duke.cs.osprey.ematrix.EnergyMatrix"><code class="xref java java-ref docutils literal"><span class="pre">EnergyMatrix</span></code></a>) – The energy matrix that defines conformation scores</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.gmec.ConfAnalyzer.html" title="edu.duke.cs.osprey.gmec.ConfAnalyzer"><code class="xref java java-ref docutils literal"><span class="pre">ConfAnalyzer</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="osprey.SequenceAnalyzer">
<code class="descclassname">osprey.</code><code class="descname">SequenceAnalyzer</code><span class="sig-paren">(</span><em>proteinConfSpace</em>, <em>ligandConfSpace</em>, <em>complexConfSpace</em>, <em>ecalc</em>, <em>confEcalcFactory</em>, <em>astarFactory</em>, <em>energyMatrixCachePattern=None</em>, <em>confDBPattern=&lt;osprey.UseJavaDefault instance&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#osprey.SequenceAnalyzer" title="Permalink to this definition">¶</a></dt>
<dd><p>Shows information about a single sequence.</p>
<p>For examples using the sequence analyzer, see examples/python.KStar/analyzeSequence.py in your Osprey distribution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>proteinConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing just the protein strand</li>
<li><strong>ligandConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing just the ligand strand</li>
<li><strong>complexConfSpace</strong> (<a class="reference internal" href="api.confspace.SimpleConfSpace.html" title="edu.duke.cs.osprey.confspace.SimpleConfSpace"><code class="xref java java-ref docutils literal"><span class="pre">SimpleConfSpace</span></code></a>) – A configuration space containing both the protein and ligand strands</li>
<li><strong>ecalc</strong> (<a class="reference internal" href="api.energy.EnergyCalculator.html" title="edu.duke.cs.osprey.energy.EnergyCalculator"><code class="xref java java-ref docutils literal"><span class="pre">EnergyCalculator</span></code></a>) – Calculates the energy for a molecule</li>
<li><strong>confEcalcFactory</strong> (<a class="reference internal" href="api.kstar.KStar.ConfEnergyCalculatorFactory.html" title="edu.duke.cs.osprey.kstar.KStar$ConfEnergyCalculatorFactory"><code class="xref java java-ref docutils literal"><span class="pre">ConfEnergyCalculatorFactory</span></code></a>) – A function that makes a ConfEnergyCalculator with the desired options</li>
<li><strong>astarFactory</strong> (<a class="reference internal" href="api.kstar.KStar.ConfSearchFactory.html" title="edu.duke.cs.osprey.kstar.KStar$ConfSearchFactory"><code class="xref java java-ref docutils literal"><span class="pre">ConfSearchFactory</span></code></a>) – A function that makes a ConfSearchFactory (e.g, A* search) with the desired options</li>
<li><strong>energyMatrixCachePattern</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – <p>Pattern of the filename to cache energy matrices.</p>
<p>K*-type algorithms must calculate multiple energy matrices.
By default, these energy matrices are not cached between runs.
To cache energy matrices between runs, supply a pattern such as:</p>
<p>”theFolder/emat.*.dat”</p>
<p>The * in the pattern is a wildcard character that will be replaced with
each type of energy matrix used by the K*-type algorithm.</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="api.kstar.SequenceAnalyzer.html" title="edu.duke.cs.osprey.kstar.SequenceAnalyzer"><code class="xref java java-ref docutils literal"><span class="pre">SequenceAnalyzer</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
            
  <div class="footer-relations">
    
      <div class="pull-left">
        <a class="btn btn-default" href="tutorial.html" title="previous chapter (use the left arrow)">Tutorial</a>
      </div>
    
    </div>
    <div class="clearer"></div>
  
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial.html" title="Tutorial"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Osprey 3.0-beta1 documentation</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2017, Donald Lab, Department of Computer Science, Duke University. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>